File: src/auth/device-flow.ts
import {
	credentialsFromTokenResponse,
	requireOAuthUnconfigured,
	storeOAuthCredentials,
} from "./credentials";
import { openJson, sealJson } from "./envelope";
import {
	DEVICE_VERIFICATION_URL,
	pollDeviceAuthorizationToken,
	requestDeviceAuthorization,
} from "./oauth-provider";
import { ApiError } from "../shared/api-error";
import { isRecord, numberField, stringField } from "../shared/json";

const DEVICE_STATE_PURPOSE = "codex-worker/device-state/v1";
const DEVICE_LIFETIME_MS = 15 * 60 * 1000;
const DEFAULT_POLL_INTERVAL_SECONDS = 5;

type OAuthEnv = Pick<Env, "AUTH_KV" | "OAUTH_MASTER_KEY">;

interface DeviceState {
	version: 1;
	deviceAuthId: string;
	userCode: string;
	expiresAt: number;
	interval: number;
}

export interface DeviceAuthorization {
	verificationUri: string;
	userCode: string;
	expiresIn: number;
	interval: number;
	state: string;
}

export type DevicePollResult =
	| { status: "pending"; retryAfter: number }
	| { status: "stored" };

export async function startDeviceAuthorization(
	env: OAuthEnv,
	signal?: AbortSignal,
): Promise<DeviceAuthorization> {
	await requireOAuthUnconfigured(env);
	const providerAuthorization = await requestDeviceAuthorization(signal);
	const interval = pollInterval(providerAuthorization.interval);
	const expiresAt = Date.now() + DEVICE_LIFETIME_MS;
	let state: string;
	try {
		state = await sealJson(
			{
				version: 1,
				deviceAuthId: providerAuthorization.deviceAuthId,
				userCode: providerAuthorization.userCode,
				expiresAt,
				interval,
			} satisfies DeviceState,
			env.OAUTH_MASTER_KEY,
			DEVICE_STATE_PURPOSE,
		);
	} catch {
		throw new ApiError(
			500,
			"Unable to create a device authorization session.",
			"configuration_error",
			"device_session_unavailable",
		);
	}

	return {
		verificationUri: DEVICE_VERIFICATION_URL,
		userCode: providerAuthorization.userCode,
		expiresIn: Math.floor(DEVICE_LIFETIME_MS / 1000),
		interval,
		state,
	};
}

export async function pollDeviceAuthorization(
	env: OAuthEnv,
	sealedState: string,
	signal?: AbortSignal,
): Promise<DevicePollResult> {
	await requireOAuthUnconfigured(env);
	const state = await deviceState(sealedState, env.OAUTH_MASTER_KEY);
	if (state.expiresAt <= Date.now()) {
		throw new ApiError(
			410,
			"The device authorization session has expired.",
			"invalid_request_error",
			"device_session_expired",
		);
	}

	const result = await pollDeviceAuthorizationToken(
		state.deviceAuthId,
		state.userCode,
		signal,
	);
	if (result.status === "pending") {
		return { status: "pending", retryAfter: state.interval };
	}

	const credentials = credentialsFromTokenResponse(
		result.tokenPayload,
		undefined,
	);
	await requireOAuthUnconfigured(env);
	await storeOAuthCredentials(env, credentials);
	return { status: "stored" };
}

async function deviceState(
	sealedState: string,
	masterKey: string,
): Promise<DeviceState> {
	let value: unknown;
	try {
		value = await openJson(sealedState, masterKey, DEVICE_STATE_PURPOSE);
	} catch {
		throw invalidDeviceState();
	}
	if (!isRecord(value)) throw invalidDeviceState();
	const deviceAuthId = stringField(value, "deviceAuthId");
	const userCode = stringField(value, "userCode");
	const expiresAt = numberField(value, "expiresAt");
	const interval = numberField(value, "interval");
	if (
		value.version !== 1 ||
		!deviceAuthId ||
		!userCode ||
		expiresAt === undefined ||
		interval === undefined ||
		interval < 1
	) {
		throw invalidDeviceState();
	}
	return { version: 1, deviceAuthId, userCode, expiresAt, interval };
}

function pollInterval(value: unknown): number {
	const parsed =
		typeof value === "number"
			? value
			: typeof value === "string"
				? Number.parseInt(value, 10)
				: Number.NaN;
	return Number.isInteger(parsed) && parsed > 0 && parsed <= 60
		? parsed
		: DEFAULT_POLL_INTERVAL_SECONDS;
}

function invalidDeviceState(): ApiError {
	return new ApiError(
		400,
		"The device authorization session is invalid.",
		"invalid_request_error",
		"invalid_device_session",
	);
}

File: worker-rs/src/auth/device_flow.rs
//! Encrypted device-authorization session orchestration.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::core::{ApiError, AppResult};

use super::{
    StoredOAuthCredentials, credentials_from_token_response,
    oauth_ports::{OAuthClock, OAuthCredentialsStore},
    oauth_provider::{DEVICE_VERIFICATION_URL, OAuthProvider, ProviderDevicePollResult},
    open_json, seal_json,
};

const DEVICE_STATE_PURPOSE: &str = "codex-worker/device-state/v1";
pub const DEVICE_LIFETIME_MS: i64 = 15 * 60 * 1_000;
pub const DEFAULT_POLL_INTERVAL_SECONDS: u64 = 5;

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeviceAuthorization {
    pub verification_uri: String,
    pub user_code: String,
    pub expires_in: u64,
    pub interval: u64,
    pub state: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum DevicePollResult {
    Pending { retry_after: u64 },
    Stored { credentials: StoredOAuthCredentials },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct DeviceState {
    version: u8,
    device_auth_id: String,
    user_code: String,
    expires_at: i64,
    interval: u64,
}

pub struct DeviceAuthorizationService<'a> {
    credentials: &'a dyn OAuthCredentialsStore,
    provider: &'a OAuthProvider<'a>,
    clock: &'a dyn OAuthClock,
    master_key: &'a str,
}

impl<'a> DeviceAuthorizationService<'a> {
    pub fn new(
        credentials: &'a dyn OAuthCredentialsStore,
        provider: &'a OAuthProvider<'a>,
        clock: &'a dyn OAuthClock,
        master_key: &'a str,
    ) -> Self {
        Self {
            credentials,
            provider,
            clock,
            master_key,
        }
    }

    pub async fn start(&self) -> AppResult<DeviceAuthorization> {
        self.credentials.require_unconfigured().await?;
        let authorization = self.provider.request_device_authorization().await?;
        let interval = poll_interval(authorization.interval.as_ref());
        let expires_at = self.clock.now_ms().await.saturating_add(DEVICE_LIFETIME_MS);
        let state = DeviceState {
            version: 1,
            device_auth_id: authorization.device_auth_id,
            user_code: authorization.user_code.clone(),
            expires_at,
            interval,
        };
        let state = serde_json::to_value(state)
            .ok()
            .and_then(|value| seal_json(&value, self.master_key, DEVICE_STATE_PURPOSE).ok())
            .ok_or_else(device_session_unavailable)?;

        Ok(DeviceAuthorization {
            verification_uri: DEVICE_VERIFICATION_URL.into(),
            user_code: authorization.user_code,
            expires_in: (DEVICE_LIFETIME_MS / 1_000) as u64,
            interval,
            state,
        })
    }

    pub async fn poll(&self, sealed_state: &str) -> AppResult<DevicePollResult> {
        self.credentials.require_unconfigured().await?;
        let state = device_state(sealed_state, self.master_key)?;
        if state.expires_at <= self.clock.now_ms().await {
            return Err(
                ApiError::new(410, "The device authorization session has expired.")
                    .with_kind("invalid_request_error")
                    .with_code("device_session_expired"),
            );
        }

        match self
            .provider
            .poll_device_authorization_token(&state.device_auth_id, &state.user_code)
            .await?
        {
            ProviderDevicePollResult::Pending => Ok(DevicePollResult::Pending {
                retry_after: state.interval,
            }),
            ProviderDevicePollResult::Authorized { token_payload } => {
                let credentials = credentials_from_token_response(
                    &token_payload,
                    None,
                    self.clock.now_ms().await,
                )?;
                self.credentials.require_unconfigured().await?;
                self.credentials.store(&credentials).await?;
                Ok(DevicePollResult::Stored { credentials })
            }
        }
    }
}

fn device_state(sealed_state: &str, master_key: &str) -> AppResult<DeviceState> {
    let value = open_json(sealed_state, master_key, DEVICE_STATE_PURPOSE)
        .map_err(|_| invalid_device_state())?;
    let state: DeviceState = serde_json::from_value(value).map_err(|_| invalid_device_state())?;
    if state.version != 1
        || state.device_auth_id.is_empty()
        || state.user_code.is_empty()
        || state.interval < 1
    {
        return Err(invalid_device_state());
    }
    Ok(state)
}

fn poll_interval(value: Option<&Value>) -> u64 {
    let parsed = match value {
        Some(Value::Number(value)) => value
            .as_f64()
            .filter(|value| value.is_finite() && value.fract() == 0.0)
            .and_then(|value| u64::try_from(value as i128).ok()),
        Some(Value::String(value)) => parse_decimal_prefix(value),
        _ => None,
    };
    parsed
        .filter(|value| (1..=60).contains(value))
        .unwrap_or(DEFAULT_POLL_INTERVAL_SECONDS)
}

fn parse_decimal_prefix(value: &str) -> Option<u64> {
    let value = value.trim_start();
    let (negative, value) = match value.as_bytes().first() {
        Some(b'+') => (false, &value[1..]),
        Some(b'-') => (true, &value[1..]),
        _ => (false, value),
    };
    if negative {
        return None;
    }
    let digits = value.bytes().take_while(u8::is_ascii_digit).count();
    if digits == 0 {
        return None;
    }
    value[..digits].parse().ok()
}

fn device_session_unavailable() -> ApiError {
    ApiError::new(500, "Unable to create a device authorization session.")
        .with_kind("configuration_error")
        .with_code("device_session_unavailable")
}

fn invalid_device_state() -> ApiError {
    ApiError::new(400, "The device authorization session is invalid.")
        .with_kind("invalid_request_error")
        .with_code("invalid_device_session")
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::*;

    #[test]
    fn matches_javascript_poll_interval_parsing() {
        assert_eq!(poll_interval(Some(&json!(1))), 1);
        assert_eq!(poll_interval(Some(&json!("  +12seconds"))), 12);
        assert_eq!(poll_interval(Some(&json!(1.5))), 5);
        assert_eq!(poll_interval(Some(&json!("61"))), 5);
        assert_eq!(poll_interval(None), 5);
    }
}
